pyro/infer/autoguide/guides.py

```diff
@@ -1482,3 +1506,350 @@ def median(self, *args, **kwargs):
+        encoder_class
+            Class for defining encoder network.
+        encoder_kwargs
+            Keyword arguments for encoder_class.
+        create_plates
+            Function for creating plates
+        single_encoder
+            Use single encoder for all variables (True) or one encoder per variable (False).
+        """
+
+        super().__init__(model, create_plates=create_plates)
+        self.amortised_plate_sites = amortised_plate_sites
+        self.single_encoder = single_encoder
+
+        self.softplus = SoftplusTransform()
+
+        encoder_kwargs = encoder_kwargs if isinstance(encoder_kwargs, dict) else dict()
+        encoder_kwargs["n_hidden"] = n_hidden
+        self.encoder_kwargs = encoder_kwargs
+
+        self.n_in = n_in
+        self.n_out = (
+            np.sum(
+                [
+                    np.sum(amortised_plate_sites["sites"][k])
+                    for k in amortised_plate_sites["sites"].keys()
+                ]
+            )
+            * 2
+        )
+        self.n_hidden = n_hidden
+        self.encoder_class = encoder_class
+        if self.single_encoder:
+            # create a single encoder NN
+            self.encoder = encoder_class(
+                n_in=self.n_in, n_out=self.n_hidden, **self.encoder_kwargs
+            )
+
+        self.init_param_scale = init_param_scale
+        self.data_transform = data_transform
+
+    def _setup_prototype(self, *args, **kwargs):
+
+        super()._setup_prototype(*args, **kwargs)
+
+        self._event_dims = {}
+        self._cond_indep_stacks = {}
+        self.hidden2locs = PyroModule()
+        self.hidden2scales = PyroModule()
+
+        if not self.single_encoder:
+            # create module class for collecting multiple encoder NN
+            self.encoder = PyroModule()
+
+        # Initialize guide params
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            # Collect unconstrained event_dims, which may differ from constrained event_dims.
+            with helpful_support_errors(site):
+                init_loc = (
+                    biject_to(site["fn"].support).inv(site["value"].detach()).detach()
+                )
+            event_dim = site["fn"].event_dim + init_loc.dim() - site["value"].dim()
+            self._event_dims[name] = event_dim
+
+            # Collect independence contexts.
+            self._cond_indep_stacks[name] = site["cond_indep_stack"]
+
+            # add linear layer for locs and scales
+            param_dim = (self.n_hidden, self.amortised_plate_sites["sites"][name])
+            init_param = torch.normal(
+                torch.full(size=param_dim, fill_value=0.0, device=site["value"].device),
+                torch.full(
+                    size=param_dim,
+                    fill_value=(1 * self.init_param_scale) / np.sqrt(self.n_hidden),
+                    device=site["value"].device,
+                ),
+            )
```

Comment:
If I use torch.normal rather than numpy.random.normal, I get this warning:

```
/nfs/team205/vk7/sanger_projects/BayraktarLab/scvi-tools/scvi/external/cell2location/autoguide.py:218: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  init_param, device=site["value"].device, requires_grad=True
```

Reply:
I also get different results after training the model.

Reply:
Numpy alternative

```
init_param = np.random.normal(
                np.zeros(param_dim),
                (np.ones(param_dim) * self.init_param_scale) / np.sqrt(self.n_hidden),
            ).astype("float32")
```